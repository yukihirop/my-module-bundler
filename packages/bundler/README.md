# My Module Bundler

Made for understanding module bundler

## Usage

```bash
yarn dev <input file> <output file>
```

## Inspire

This module bundler uses the following code as a reference.

- [ronami/minipack](https://github.com/ronami/minipack)
- [hiroppy/the-sample-of-module-bundler](https://github.com/hiroppy/the-sample-of-module-bundler)

## Reference

- [jamiebuilds/babel-handbook(ja)](https://github.com/jamiebuilds/babel-handbook/blob/master/translations/ja/plugin-handbook.md)
- https://github.com/TypeStrong/ts-node/issues/537
- https://stackoverflow.com/questions/52092739/upgrade-to-babel-7-cannot-read-property-bindings-of-null
