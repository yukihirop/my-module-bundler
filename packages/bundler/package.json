{
  "name": "bundler",
  "version": "1.0.0",
  "main": "index.js",
  "license": "MIT",
  "scripts": {
    "dev": "ts-node src/index.ts",
    "fmt": "prettier --write src/**/*.ts"
  },
  "devDependencies": {
    "@babel/preset-env": "^7.10.2",
    "@types/babel__core": "^7.1.8",
    "@types/babel__preset-env": "^7.9.0",
    "@types/babel__traverse": "^7.0.12",
    "@types/babylon": "^6.16.5",
    "@types/jest": "^26.0.0",
    "@types/js-beautify": "^1.11.0",
    "@types/node": "^14.0.13",
    "babel-core": "^6.26.3",
    "babel-traverse": "^6.26.0",
    "babylon": "^6.18.0",
    "js-beautify": "^1.11.0",
    "ts-jest": "^26.1.0",
    "ts-node": "^8.10.2",
    "typescript": "^3.9.5"
  }
}
