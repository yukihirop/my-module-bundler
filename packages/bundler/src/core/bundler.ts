'use strict';

import fs from 'fs';
import path from 'path';
import { default as traverse } from '@babel/traverse';
import * as parser from '@babel/parser';
import { transformFromAstSync } from '@babel/core';
import { mainTemplate, moduleTemlate } from './template';
import { js as beautify } from 'js-beautify';
// TypeError: unknown: Cannot read property 'bindings' of null
// https://stackoverflow.com/questions/52092739/upgrade-to-babel-7-cannot-read-property-bindings-of-null
import env from '@babel/preset-env';
import { AssetType, GraphType, WriteParamsType, OptionsType } from './types';

// TODO: Use Class
class Bunlder {
  public moduleId: number;
  public options: OptionsType;

  constructor(options: OptionsType) {
    this.moduleId = 0;
    this.options = options;
  }

  /**
   * We create a function that will accept a path to a file, read
   * it's contents, and extract its dependencies.
   *
   */
  private createAsset(filename: string): AssetType {
    const content = fs.readFileSync(filename, 'utf-8');
    const ast: any = parser.parse(content, {
      sourceType: 'module',
    });

    // This array will hold the relative paths of modules this module depends on.
    const dependencies = [];

    traverse(ast, {
      ImportDeclaration({ node }) {
        // Every time we see an import into the dependencies array.
        dependencies.push(node.source.value);
      },
      ExportAllDeclaration({ node }) {
        // Every time we see an import into the dependencies array.
        dependencies.push(node.source.value);
      },
      ExportNamedDeclaration({ node }) {
        // Every time we see an import into the dependencies array.
        if (node.source) {
          dependencies.push(node.source.value);
        }
      },
    });

    // We also assign a unique identifier to this module by incrementing a simple counter.
    const id = this.moduleId++;

    const { code } = transformFromAstSync(ast, null, this.options);

    return {
      id,
      filename,
      dependencies,
      code,
    };
  }

  /**
   * Extract the dependencies of the entry file
   */
  private createGraph(entry: string): GraphType {
    const mainAsset = this.createAsset(entry);
    const queue = [mainAsset];

    for (const asset of queue) {
      // assets has alist of relative paths to he modules it depends on.
      asset.mapping = {};
      const dirname = path.dirname(asset.filename);

      asset.dependencies.forEach((relativePath) => {
        const absolutePath = path.join(dirname, relativePath);
        // Parse the asset, read its content, and extract its dependencies.
        const child = this.createAsset(absolutePath);
        // It's essential for us to know `asset` depends on `child`
        asset.mapping[relativePath] = child.id;

        queue.push(child);
      });
    }

    return queue;
  }

  /**
   * Define a function that will use our graph and return a bundle that
   * We can run in the browser
   *
   * Our bundle will have just one self-invoking function:
   *
   * (function(){})()
   * @access private
   */
  private bundle(graph: GraphType): string {
    let modules = [];

    // Add a string of this format: `key: value,`.
    graph.forEach((mod) => {
      modules.push(moduleTemlate(mod));
    });

    // Finaly, we implement the body of the self-invoking function (IFEE)
    return mainTemplate(modules);
  }

  public async write({ input, output }: WriteParamsType) {
    const graph = this.createGraph(input),
      result = this.bundle(graph),
      formatted = beautify(result, {
        indent_size: 2,
        space_in_empty_paren: true,
      });

    await fs.promises.writeFile(output, formatted);
  }
}

export default Bunlder;
