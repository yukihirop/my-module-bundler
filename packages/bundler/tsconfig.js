const { default: lib } = require("babel-preset-env");

module.exports = {
  include: ["src"],
  compilerOptions: {
    outDir: "lib",
    rootDir: "src",
    declaration: true,
    module: "esnext"
  }
}
