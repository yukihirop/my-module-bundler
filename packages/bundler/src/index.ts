'use strict';

import fs from 'fs';
import transformArrowFunctions from '@yukihirop/plugin-transform-arrow-functions';
import transformModulesCommonjs from '@yukihirop/plugin-transform-modules-commonjs';

import Bundler from './core/bundler';
import { OptionsType } from './core/types';

const isExistFile = (file) => {
  try {
    fs.statSync(file);
    return true;
  } catch (err) {
    if (err.code === 'ENOENT') return false;
  }
};

const currentPath = fs.realpathSync('./');
const settingPath = process.env.MMBRC_PATH ? process.env.MMBRC_PATH : `${currentPath}/.mmbrc`;
const options: OptionsType | undefined = isExistFile(settingPath)
  ? require(settingPath)
  : undefined;

const input = process.argv[2],
  output = process.argv[3],
  defaultOptions = {
    plugins: [
      [transformArrowFunctions, {}],
      [transformModulesCommonjs, {}],
    ],
  } as OptionsType;
const bundler = new Bundler(options || defaultOptions);
bundler.write({ input, output });
