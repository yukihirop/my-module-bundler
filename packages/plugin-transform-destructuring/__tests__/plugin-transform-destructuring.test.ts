'use strict';

import { promises } from 'fs';
import { runInContext, createContext } from 'vm';
import { join } from 'path';

import { Bundler, types as t } from '@yukihirop/bundler';
import transformDestructuring from '../src';

const { readFile, mkdir } = promises;
const fixtureBasePath = join(__dirname, 'fixtures')
  , outputBasePath = join(__dirname, 'output')
  , ENTRY_FILE = 'entry.js'
  , BUNDLE_FILE = 'bundle.js'
  , defaultOpts = {
    "plugins": [
      [transformDestructuring, {}],
    ]
  } as t.OptionsType;

type TestBuildOptionsType = {
  opts?: t.OptionsType
  type?: string
}

async function build(fixturePath: string, outputPath: string, options: TestBuildOptionsType) {
  const { opts, type } = options;
  await new Bundler(opts || defaultOpts).write({
    input: join(fixturePath, type || "", ENTRY_FILE),
    output: join(outputPath, type || "", BUNDLE_FILE)
  })
}

async function runGeneratedCodeInVM(outputPath: string, type?: string) {
  const code = await readFile(join(outputPath, type || "", BUNDLE_FILE), 'utf-8'),
    sandbox = { console, process },
    ctx = createContext({ sandbox });

  runInContext(code, ctx);
}

async function createDir(base: string, dir?: string) {
  await mkdir(join(base, dir || ""), { recursive: true });
}

beforeEach(() => {
  jest.spyOn(console, 'log').mockImplementation()
  jest.spyOn(console, 'warn').mockImplementation()
});

afterEach(() => {
  jest.clearAllMocks();
});

describe('destructuring', () => {
  const type = 'destructuring'
  const fixturePath = join(fixtureBasePath, type)
  const outputPath = join(outputBasePath, type)

  beforeAll(async () => {
    await createDir(outputPath, '');
  })

  const dirs = [
    'array-basic',
    'array-rest-basic',
    'array-rest-nested',
    'array-overflow',
    'array-self-reference',
    'array-assignment-function-block',
    'object-basic',
    'object-rest-basic',
    'object-rest-nested',
    'object-overflow',
    'object-assignment-function-block',
    'call-expression-array-basic',
    'call-expression-array-helper-rename'
  ]

  for (const dir of dirs) {
    test(dir, async () => {
      await createDir(outputPath, dir);
      await build(fixturePath, outputPath, { type: dir });
      await runGeneratedCodeInVM(outputPath, dir);

      const code = await readFile(join(outputPath, dir, BUNDLE_FILE), 'utf-8')
      expect(code).toMatchSnapshot();
    });
  }
})
