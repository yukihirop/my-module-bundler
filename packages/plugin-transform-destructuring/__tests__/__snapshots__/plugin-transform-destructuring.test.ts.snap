// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`destructuring array-assignment-function-block 1`] = `
"(function(modules) {
  function require(id) {
    const [fn, mapping] = modules[id];

    function localRequire(name) {
      return require(mapping[name]);
    }

    const module = {
      exports: {}
    };

    fn(localRequire, module, module.exports);

    return module.exports;
  }

  require(0)
})({
  0: [
    function(require, module, exports) {
      function c() {
        var a = 1,
          b = 2;
      }

      ;
    },
    {},
  ]
})"
`;

exports[`destructuring array-basic 1`] = `
"(function(modules) {
  function require(id) {
    const [fn, mapping] = modules[id];

    function localRequire(name) {
      return require(mapping[name]);
    }

    const module = {
      exports: {}
    };

    fn(localRequire, module, module.exports);

    return module.exports;
  }

  require(0)
})({
  0: [
    function(require, module, exports) {
      var a = 1,
        b = '2',
        c = true;
      const d = 1,
        e = '2',
        f = true;
      let g = 1,
        h = '2',
        i = true;
    },
    {},
  ]
})"
`;

exports[`destructuring array-overflow 1`] = `
"(function(modules) {
  function require(id) {
    const [fn, mapping] = modules[id];

    function localRequire(name) {
      return require(mapping[name]);
    }

    const module = {
      exports: {}
    };

    fn(localRequire, module, module.exports);

    return module.exports;
  }

  require(0)
})({
  0: [
    function(require, module, exports) {
      var _ref = [1, '2', true, null, undefined, function() {}, Error, WebAssembly],
        a = _ref[0],
        b = _ref[1],
        c = _ref[2];
      var _ref2 = [1, '2', true, null, undefined, function() {}, Error, WebAssembly],
        d = _ref2[0],
        e = _ref2[1],
        f = _ref2[2];
      var _ref3 = [1, '2', true, null, undefined, function() {}, Error, WebAssembly],
        g = _ref3[0],
        h = _ref3[1],
        i = _ref3[2];
    },
    {},
  ]
})"
`;

exports[`destructuring array-rest-basic 1`] = `
"(function(modules) {
  function require(id) {
    const [fn, mapping] = modules[id];

    function localRequire(name) {
      return require(mapping[name]);
    }

    const module = {
      exports: {}
    };

    fn(localRequire, module, module.exports);

    return module.exports;
  }

  require(0)
})({
  0: [
    function(require, module, exports) {
      var a = 1,
        b = '2',
        c = [true, null, undefined, function() {}, Error, WebAssembly];
      const d = 1,
        e = '2',
        f = [true, null, undefined, function() {}, Error, WebAssembly];
      let g = 1,
        h = '2',
        i = [true, null, undefined, function() {}, Error, WebAssembly];
    },
    {},
  ]
})"
`;

exports[`destructuring array-rest-nested 1`] = `
"(function(modules) {
  function require(id) {
    const [fn, mapping] = modules[id];

    function localRequire(name) {
      return require(mapping[name]);
    }

    const module = {
      exports: {}
    };

    fn(localRequire, module, module.exports);

    return module.exports;
  }

  require(0)
})({
  0: [
    function(require, module, exports) {
      var a = 1,
        b = '2',
        c = [true, null, undefined],
        d = [function() {}, Error, WebAssembly];
    },
    {},
  ]
})"
`;

exports[`destructuring array-self-reference 1`] = `
"(function(modules) {
  function require(id) {
    const [fn, mapping] = modules[id];

    function localRequire(name) {
      return require(mapping[name]);
    }

    const module = {
      exports: {}
    };

    fn(localRequire, module, module.exports);

    return module.exports;
  }

  require(0)
})({
  0: [
    function(require, module, exports) {
      var d;
      var _ref = [1, a, b, d],
        a = _ref[0],
        b = _ref[1],
        c = _ref[2];
      var a = 1,
        b = \\"a\\",
        c = \\"b\\";
    },
    {},
  ]
})"
`;

exports[`destructuring call-expression-array-basic 1`] = `
"(function(modules) {
  function require(id) {
    const [fn, mapping] = modules[id];

    function localRequire(name) {
      return require(mapping[name]);
    }

    const module = {
      exports: {}
    };

    fn(localRequire, module, module.exports);

    return module.exports;
  }

  require(0)
})({
  0: [
    function(require, module, exports) {
      function _udf_slicedToArray(arr, i) {
        return _udf_arrayWithHoles(arr) || _udf_iterableToArrayLimit(arr, i) || _udf_unsupportedIterableToArray(arr, i) || _udf_nonIterableRest();
      }

      function _udf_nonIterableRest() {
        throw new TypeError(\\"Invalid attempt to destructure non-iterable instance.\\\\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method.\\");
      }

      function _udf_unsupportedIterableToArray(o, minLen) {
        if (!o) return;
        if (typeof o === \\"string\\") return _udf_arrayLikeToArray(o, minLen);
        var n = Object.prototype.toString.call(o).slice(8, -1);
        if (n === \\"Object\\" && o.constructor) n = o.constructor.name;
        if (n === \\"Map\\" || n === \\"Set\\") return Array.from(o);
        if (n === \\"Arguments\\" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(n)) return _udf_arrayLikeToArray(o, minLen);
      }

      function _udf_arrayLikeToArray(arr, len) {
        if (len == null || len > arr.length) len = arr.length;
        for (var i = 0, arr2 = new Array(len); i < len; i++) {
          arr2[i] = arr[i];
        }
        return arr2;
      }

      function _udf_iterableToArrayLimit(arr, i) {
        if (typeof Symbol === \\"undefined\\" || !(Symbol.iterator in Object(arr))) return;
        var _arr = [];
        var _n = true;
        var _d = false;
        var _e = undefined;
        try {
          for (var _i = arr[Symbol.iterator](), _s; !(_n = (_s = _i.next()).done); _n = true) {
            _arr.push(_s.value);
            if (i && _arr.length === i) break;
          }
        } catch (err) {
          _d = true;
          _e = err;
        } finally {
          try {
            if (!_n && _i[\\"return\\"] != null) _i[\\"return\\"]();
          } finally {
            if (_d) throw _e;
          }
        }
        return _arr;
      }

      function _udf_arrayWithHoles(arr) {
        if (Array.isArray(arr)) return arr;
      }

      function f() {
        return [1, 2];
      }

      ;

      var _f = f(),
        _f2 = _udf_slicedToArray(_f, 2),
        a = _f2[0],
        b = _f2[1];

      function g() {
        return [1, 2];
      }

      ;

      var _g = g(),
        _g2 = _udf_slicedToArray(_g, 2),
        c = _g2[0],
        d = _g2[1];
    },
    {},
  ]
})"
`;

exports[`destructuring call-expression-array-helper-rename 1`] = `
"(function(modules) {
  function require(id) {
    const [fn, mapping] = modules[id];

    function localRequire(name) {
      return require(mapping[name]);
    }

    const module = {
      exports: {}
    };

    fn(localRequire, module, module.exports);

    return module.exports;
  }

  require(0)
})({
  0: [
    function(require, module, exports) {
      function _udf_slicedToArray(arr, i) {
        return _udf_arrayWithHoles(arr) || _udf_iterableToArrayLimit(arr, i) || _udf_unsupportedIterableToArray(arr, i) || _udf_nonIterableRest();
      }

      function _udf_nonIterableRest() {
        throw new TypeError(\\"Invalid attempt to destructure non-iterable instance.\\\\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method.\\");
      }

      function _udf_unsupportedIterableToArray(o, minLen) {
        if (!o) return;
        if (typeof o === \\"string\\") return _udf_arrayLikeToArray(o, minLen);
        var n = Object.prototype.toString.call(o).slice(8, -1);
        if (n === \\"Object\\" && o.constructor) n = o.constructor.name;
        if (n === \\"Map\\" || n === \\"Set\\") return Array.from(o);
        if (n === \\"Arguments\\" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(n)) return _udf_arrayLikeToArray(o, minLen);
      }

      function _udf_arrayLikeToArray(arr, len) {
        if (len == null || len > arr.length) len = arr.length;
        for (var i = 0, arr2 = new Array(len); i < len; i++) {
          arr2[i] = arr[i];
        }
        return arr2;
      }

      function _udf_iterableToArrayLimit(arr, i) {
        if (typeof Symbol === \\"undefined\\" || !(Symbol.iterator in Object(arr))) return;
        var _arr = [];
        var _n = true;
        var _d = false;
        var _e = undefined;
        try {
          for (var _i = arr[Symbol.iterator](), _s; !(_n = (_s = _i.next()).done); _n = true) {
            _arr.push(_s.value);
            if (i && _arr.length === i) break;
          }
        } catch (err) {
          _d = true;
          _e = err;
        } finally {
          try {
            if (!_n && _i[\\"return\\"] != null) _i[\\"return\\"]();
          } finally {
            if (_d) throw _e;
          }
        }
        return _arr;
      }

      function _udf_arrayWithHoles(arr) {
        if (Array.isArray(arr)) return arr;
      }

      function _slicedToArray() {
        return [1, 2];
      }

      ;

      function _nonIterableRest() {}

      function _unsupportedIterableToArray(o, minLen) {}

      function _arrayLikeToArray(arr, len) {}

      function _iterableToArrayLimit(arr, i) {}

      function _arrayWithHoles(arr) {}

      var _slicedToArray2 = _slicedToArray(),
        _slicedToArray3 = _udf_slicedToArray(_slicedToArray2, 2),
        a = _slicedToArray3[0],
        b = _slicedToArray3[1];
    },
    {},
  ]
})"
`;

exports[`destructuring call-expression-array-self-reference 1`] = `
"(function(modules) {
  function require(id) {
    const [fn, mapping] = modules[id];

    function localRequire(name) {
      return require(mapping[name]);
    }

    const module = {
      exports: {}
    };

    fn(localRequire, module, module.exports);

    return module.exports;
  }

  require(0)
})({
  0: [
    function(require, module, exports) {
      function _udf_slicedToArray(arr, i) {
        return _udf_arrayWithHoles(arr) || _udf_iterableToArrayLimit(arr, i) || _udf_unsupportedIterableToArray(arr, i) || _udf_nonIterableRest();
      }

      function _udf_nonIterableRest() {
        throw new TypeError(\\"Invalid attempt to destructure non-iterable instance.\\\\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method.\\");
      }

      function _udf_unsupportedIterableToArray(o, minLen) {
        if (!o) return;
        if (typeof o === \\"string\\") return _udf_arrayLikeToArray(o, minLen);
        var n = Object.prototype.toString.call(o).slice(8, -1);
        if (n === \\"Object\\" && o.constructor) n = o.constructor.name;
        if (n === \\"Map\\" || n === \\"Set\\") return Array.from(o);
        if (n === \\"Arguments\\" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(n)) return _udf_arrayLikeToArray(o, minLen);
      }

      function _udf_arrayLikeToArray(arr, len) {
        if (len == null || len > arr.length) len = arr.length;
        for (var i = 0, arr2 = new Array(len); i < len; i++) {
          arr2[i] = arr[i];
        }
        return arr2;
      }

      function _udf_iterableToArrayLimit(arr, i) {
        if (typeof Symbol === \\"undefined\\" || !(Symbol.iterator in Object(arr))) return;
        var _arr = [];
        var _n = true;
        var _d = false;
        var _e = undefined;
        try {
          for (var _i = arr[Symbol.iterator](), _s; !(_n = (_s = _i.next()).done); _n = true) {
            _arr.push(_s.value);
            if (i && _arr.length === i) break;
          }
        } catch (err) {
          _d = true;
          _e = err;
        } finally {
          try {
            if (!_n && _i[\\"return\\"] != null) _i[\\"return\\"]();
          } finally {
            if (_d) throw _e;
          }
        }
        return _arr;
      }

      function _udf_arrayWithHoles(arr) {
        if (Array.isArray(arr)) return arr;
      }

      function f() {
        return [c, d];
      }

      ;

      var _f = f(),
        _f2 = _udf_slicedToArray(_f, 2),
        a = _f2[0],
        b = _f2[1];

      function g() {
        return [a, b];
      }

      ;

      var _g = g(),
        _g2 = _udf_slicedToArray(_g, 2),
        c = _g2[0],
        d = _g2[1];
    },
    {},
  ]
})"
`;

exports[`destructuring object-assignment-function-block 1`] = `
"(function(modules) {
  function require(id) {
    const [fn, mapping] = modules[id];

    function localRequire(name) {
      return require(mapping[name]);
    }

    const module = {
      exports: {}
    };

    fn(localRequire, module, module.exports);

    return module.exports;
  }

  require(0)
})({
  0: [
    function(require, module, exports) {
      function c() {
        var _a$b = {
            a: 1,
            b: 2
          },
          a = _a$b.a,
          b = _a$b.b;
      }

      ;
    },
    {},
  ]
})"
`;

exports[`destructuring object-basic 1`] = `
"(function(modules) {
  function require(id) {
    const [fn, mapping] = modules[id];

    function localRequire(name) {
      return require(mapping[name]);
    }

    const module = {
      exports: {}
    };

    fn(localRequire, module, module.exports);

    return module.exports;
  }

  require(0)
})({
  0: [
    function(require, module, exports) {
      var _a$b$c = {
          a: 1,
          b: '2',
          c: true
        },
        a = _a$b$c.a,
        b = _a$b$c.b,
        c = _a$b$c.c;
      var _d$e$f = {
          d: 1,
          e: '2',
          f: true
        },
        d = _d$e$f.d,
        e = _d$e$f.e,
        f = _d$e$f.f;
      var _g$h$i = {
          g: 1,
          h: '2',
          i: true
        },
        g = _g$h$i.g,
        h = _g$h$i.h,
        i = _g$h$i.i;
    },
    {},
  ]
})"
`;

exports[`destructuring object-overflow 1`] = `
"(function(modules) {
  function require(id) {
    const [fn, mapping] = modules[id];

    function localRequire(name) {
      return require(mapping[name]);
    }

    const module = {
      exports: {}
    };

    fn(localRequire, module, module.exports);

    return module.exports;
  }

  require(0)
})({
  0: [
    function(require, module, exports) {
      var _a$b$c$d$e$f = {
          a: 1,
          b: {
            b1: '2'
          },
          c: {
            c1: true,
            c2: null,
            c3: undefined
          },
          d: {
            d1: {
              d2: function() {}
            }
          },
          e: Error,
          f: WebAssembly
        },
        a = _a$b$c$d$e$f.a,
        b = _a$b$c$d$e$f.b,
        c = _a$b$c$d$e$f.c;
      var _a$b$c$d$e$f2 = {
          a: 1,
          b: {
            b1: '2'
          },
          c: {
            c1: true,
            c2: null,
            c3: undefined
          },
          d: {
            d1: {
              d2: function() {}
            }
          },
          e: Error,
          f: WebAssembly
        },
        d = _a$b$c$d$e$f2.d,
        e = _a$b$c$d$e$f2.e,
        f = _a$b$c$d$e$f2.f;
      var _a$b$c$d$e$f3 = {
          a: 1,
          b: {
            b1: '2'
          },
          c: {
            c1: true,
            c2: null,
            c3: undefined
          },
          d: {
            d1: {
              d2: function() {}
            }
          },
          e: Error,
          f: WebAssembly
        },
        g = _a$b$c$d$e$f3.g,
        h = _a$b$c$d$e$f3.h,
        i = _a$b$c$d$e$f3.i;
    },
    {},
  ]
})"
`;

exports[`destructuring object-rest-basic 1`] = `
"(function(modules) {
  function require(id) {
    const [fn, mapping] = modules[id];

    function localRequire(name) {
      return require(mapping[name]);
    }

    const module = {
      exports: {}
    };

    fn(localRequire, module, module.exports);

    return module.exports;
  }

  require(0)
})({
  0: [
    function(require, module, exports) {
      function _udf_objectWithoutProperties(source, excluded) {
        if (source == null) return {};
        var target = _udf_objectWithoutPropertiesLoose(source, excluded);
        var key, i;
        if (Object.getOwnPropertySymbols) {
          var sourceSymbolKeys = Object.getOwnPropertySymbols(source);
          for (i = 0; i < sourceSymbolKeys.length; i++) {
            key = sourceSymbolKeys[i];
            if (excluded.indexOf(key) >= 0) continue;
            if (!Object.prototype.propertyEnumerable.call(source, key)) continue;
            target[key] = source[key];
          }
        }
        return target;
      }

      function _udf_objectWithoutPropertiesLoose(source, excluded) {
        if (source == null) return {};
        var target = {};
        var sourceKeys = Object.keys(source);
        var key, i;
        for (i = 0; i < sourceKeys.length; i++) {
          key = sourceKeys[i];
          if (excluded.indexOf(key) >= 0) continue;
          target[key] = source[key];
        }
        return target;
      }

      var _a$b$c$d$e$f$g$h = {
          a: 1,
          b: '2',
          c: true,
          d: null,
          e: undefined,
          f: function() {},
          g: Error,
          h: WebAssembly
        },
        a = _a$b$c$d$e$f$g$h.a,
        b = _a$b$c$d$e$f$g$h.b,
        x = _udf_objectWithoutProperties(_a$b$c$d$e$f$g$h, [\\"a\\", \\"b\\"]);

      var _a$b$c$d$e$f$g$h2 = {
          a: 1,
          b: '2',
          c: true,
          d: null,
          e: undefined,
          f: function() {},
          g: Error,
          h: WebAssembly
        },
        d = _a$b$c$d$e$f$g$h2.d,
        e = _a$b$c$d$e$f$g$h2.e,
        y = _udf_objectWithoutProperties(_a$b$c$d$e$f$g$h2, [\\"d\\", \\"e\\"]);

      var _a$b$c$d$e$f$g$h3 = {
          a: 1,
          b: '2',
          c: true,
          d: null,
          e: undefined,
          f: function() {},
          g: Error,
          h: WebAssembly
        },
        m = _a$b$c$d$e$f$g$h3.m,
        n = _a$b$c$d$e$f$g$h3.n,
        z = _udf_objectWithoutProperties(_a$b$c$d$e$f$g$h3, [\\"m\\", \\"n\\"]);
    },
    {},
  ]
})"
`;

exports[`destructuring object-rest-nested 1`] = `
"(function(modules) {
  function require(id) {
    const [fn, mapping] = modules[id];

    function localRequire(name) {
      return require(mapping[name]);
    }

    const module = {
      exports: {}
    };

    fn(localRequire, module, module.exports);

    return module.exports;
  }

  require(0)
})({
  0: [
    function(require, module, exports) {
      function _udf_objectWithoutProperties(source, excluded) {
        if (source == null) return {};
        var target = _udf_objectWithoutPropertiesLoose(source, excluded);
        var key, i;
        if (Object.getOwnPropertySymbols) {
          var sourceSymbolKeys = Object.getOwnPropertySymbols(source);
          for (i = 0; i < sourceSymbolKeys.length; i++) {
            key = sourceSymbolKeys[i];
            if (excluded.indexOf(key) >= 0) continue;
            if (!Object.prototype.propertyEnumerable.call(source, key)) continue;
            target[key] = source[key];
          }
        }
        return target;
      }

      function _udf_objectWithoutPropertiesLoose(source, excluded) {
        if (source == null) return {};
        var target = {};
        var sourceKeys = Object.keys(source);
        var key, i;
        for (i = 0; i < sourceKeys.length; i++) {
          key = sourceKeys[i];
          if (excluded.indexOf(key) >= 0) continue;
          target[key] = source[key];
        }
        return target;
      }

      var _a$b$c$d$e$f = {
          a: 1,
          b: {
            b1: '2'
          },
          c: {
            c1: true,
            c2: null,
            c3: undefined
          },
          d: {
            d1: {
              d2: function() {}
            }
          },
          e: Error,
          f: WebAssembly
        },
        a = _a$b$c$d$e$f.a,
        b = _a$b$c$d$e$f.b,
        c = _a$b$c$d$e$f.c,
        d = _udf_objectWithoutProperties(_a$b$c$d$e$f, [\\"a\\", \\"b\\", \\"c\\"]);

      var _a$b$c$d$e$f2 = {
          a: 1,
          b: {
            b1: '2'
          },
          c: {
            c1: true,
            c2: null,
            c3: undefined
          },
          d: {
            d1: {
              d2: function() {}
            }
          },
          e: Error,
          f: WebAssembly
        },
        e = _a$b$c$d$e$f2.e,
        f = _a$b$c$d$e$f2.f,
        g = _a$b$c$d$e$f2.g,
        h = _udf_objectWithoutProperties(_a$b$c$d$e$f2, [\\"e\\", \\"f\\", \\"g\\"]);

      var _a$b$c$d$e$f3 = {
          a: 1,
          b: {
            b1: '2'
          },
          c: {
            c1: true,
            c2: null,
            c3: undefined
          },
          d: {
            d1: {
              d2: function() {}
            }
          },
          e: Error,
          f: WebAssembly
        },
        i = _a$b$c$d$e$f3.i,
        j = _a$b$c$d$e$f3.j,
        k = _a$b$c$d$e$f3.k,
        l = _udf_objectWithoutProperties(_a$b$c$d$e$f3, [\\"i\\", \\"j\\", \\"k\\"]);
    },
    {},
  ]
})"
`;

exports[`destructuring(err) call-expression-array-init-mix 1`] = `
"(function(modules) {
  function require(id) {
    const [fn, mapping] = modules[id];

    function localRequire(name) {
      return require(mapping[name]);
    }

    const module = {
      exports: {}
    };

    fn(localRequire, module, module.exports);

    return module.exports;
  }

  require(0)
})({
  0: [
    function(require, module, exports) {
      function _udf_slicedToArray(arr, i) {
        return _udf_arrayWithHoles(arr) || _udf_iterableToArrayLimit(arr, i) || _udf_unsupportedIterableToArray(arr, i) || _udf_nonIterableRest();
      }

      function _udf_nonIterableRest() {
        throw new TypeError(\\"Invalid attempt to destructure non-iterable instance.\\\\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method.\\");
      }

      function _udf_unsupportedIterableToArray(o, minLen) {
        if (!o) return;
        if (typeof o === \\"string\\") return _udf_arrayLikeToArray(o, minLen);
        var n = Object.prototype.toString.call(o).slice(8, -1);
        if (n === \\"Object\\" && o.constructor) n = o.constructor.name;
        if (n === \\"Map\\" || n === \\"Set\\") return Array.from(o);
        if (n === \\"Arguments\\" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(n)) return _udf_arrayLikeToArray(o, minLen);
      }

      function _udf_arrayLikeToArray(arr, len) {
        if (len == null || len > arr.length) len = arr.length;
        for (var i = 0, arr2 = new Array(len); i < len; i++) {
          arr2[i] = arr[i];
        }
        return arr2;
      }

      function _udf_iterableToArrayLimit(arr, i) {
        if (typeof Symbol === \\"undefined\\" || !(Symbol.iterator in Object(arr))) return;
        var _arr = [];
        var _n = true;
        var _d = false;
        var _e = undefined;
        try {
          for (var _i = arr[Symbol.iterator](), _s; !(_n = (_s = _i.next()).done); _n = true) {
            _arr.push(_s.value);
            if (i && _arr.length === i) break;
          }
        } catch (err) {
          _d = true;
          _e = err;
        } finally {
          try {
            if (!_n && _i[\\"return\\"] != null) _i[\\"return\\"]();
          } finally {
            if (_d) throw _e;
          }
        }
        return _arr;
      }

      function _udf_arrayWithHoles(arr) {
        if (Array.isArray(arr)) return arr;
      }

      function f() {
        return [1, 2];
      }

      ;

      var _f = f(),
        _f2 = _udf_slicedToArray(_f, 2),
        a = _f2[0],
        b = _f2[1];

      function g() {
        return {
          c: 1,
          d: 1
        };
      }

      ;

      var _g = g(),
        _g2 = _udf_slicedToArray(_g, 2),
        c = _g2[0],
        d = _g2[1];
    },
    {},
  ]
})"
`;
